src/formatting.rs

```diff
@@ -27,6 +28,8 @@ mod newline_style;
 pub(crate) type SourceFile = Vec<FileRecord>;
 pub(crate) type FileRecord = (FileName, String);
 
+thread_local!(static CAN_RESET_CRATE_PARSE_ERRORS: Cell<bool> = Cell::new(false));
```

Comment:
As mentioned in the PR description, I'm not crazy about using a global here 😞 I got the idea when I saw something similar being leveraged internally in the parser:

- https://github.com/rust-lang/rust/blob/52b3d1f2eb339a2c414b7a18bdf5331308a69192/src/librustc_errors/lib.rs#L333
- https://github.com/rust-lang/rust/blob/52b3d1f2eb339a2c414b7a18bdf5331308a69192/src/librustc_errors/lib.rs#L668-L679

I briefly considered trying to hook into that `TRACKING_DIAGNOSTICS`, though it seems that to do so I'd still need a global to store the Diagnostic events as they roll through, and would then have to re-analyze them all after parsing, so went this route instead.